tests/engine/preprocessing/test_embeder.py

```diff
@@ -0,0 +1,85 @@
+# pylint: disable-all
+import pandas as pd
+import pytest
+import torch
+
+from recommender.engine.models.autoencoders import AutoEncoder
+from recommender.engine.preprocessing.embedder import Embedder
+from recommender.errors import MissingDenseTensorError, MissingOneHotTensorError
+from tests.factories.marketplace import ServiceFactory
+
+
+@pytest.fixture
+def services(mongo):
+    return [ServiceFactory(id=i) for i in range(10)]
+
+
+def test_embedder(mongo, services):
```

Comment:
I know this might be inconvenient, but this test should be split into smaller test cases, for clarity. I would avoid writing very long tests that test everything and prefer smaller, well defined test cases - to refactor later

Reply:
Godd suggestion but in this case each line is advancing previous setup, so next assertion uses changed test setup of the previous assertions. This is the only reason to have all of these assertion in one test case. If we are spliting this into multiple methods we would have to make a lot of repeated setup.

Reply:
But in this particular case it's good approach, because we have to use embedders to make scenario for other embedders tests (it is inevitable). We know exactly which functionality is failing because each assertion set is executed sequentially, so when one of them is failing you know that you have to first fix this assertion and not to fix next potentially failing assertions because after you fix the first one, next could be resolved too. If it will be not it doesn't matter, you should still fix the new one first failing assertion to make sure that it doesn't interfere with the next.

To sum it up: you will always know which part of test code has something to do with fixing when something is not working: the part from the beginning of the test to the failing assertion

Reply:
> But in this particular case it's good approach, because we have to use embedders to make scenario for other embedders tests (it is inevitable). We know exactly which functionality is failing because each assertion set is executed sequentially, so when one of them is failing you know that you have to first fix this assertion and not to fix next potentially failing assertions because after you fix the first one, next could be resolved too. If it will be not it doesn't matter, you should still fix the new one first failing assertion to make sure that it doesn't interfere with the next.
> 
> To sum it up: you will always know which part of test code has something to do with fixing when something is not working: the part from the beginning of the test to the failing assertion

This is true, and can stay that way, but you will agree that if you see that a particular, more granular test case is failing it is more clear what to fix and what is failing, than to check individual assertions) - this is also useful for checking for regression in the upcoming versions of the code

Reply:
> > But in this particular case it's good approach, because we have to use embedders to make scenario for other embedders tests (it is inevitable). We know exactly which functionality is failing because each assertion set is executed sequentially, so when one of them is failing you know that you have to first fix this assertion and not to fix next potentially failing assertions because after you fix the first one, next could be resolved too. If it will be not it doesn't matter, you should still fix the new one first failing assertion to make sure that it doesn't interfere with the next.
> > To sum it up: you will always know which part of test code has something to do with fixing when something is not working: the part from the beginning of the test to the failing assertion
> 
> This is true, and can stay that way, but you will agree that if you see that a particular, more granular test case is failing it is more clear what to fix and what is failing, than to check individual assertions) - this is also useful for checking for regression in the upcoming versions of the code

sure, I think generaly spliting it in more tests is a good idea, but still I don't know how to do it gracefully in pytest - but definitely it can be considered in the next issue.

Reply:
> Yeah, in ruby, but how the gradual setup should be implemented here in pytest?

Using fixtures, you can do literally anything (including running some code, not only returning some objects and then tearing down the setup if needed, using the 'yield' keyword)

Reply:
I agree with @wujuu that the idea behind this test is a little bit hard to follow. Refactoring or solid description/documentation can be a solution. Of course, taking under consideration tasks priority.